frontend/src/metabase/visualizations/visualizations/Heading/Heading.unit.spec.tsx

```diff
@@ -0,0 +1,123 @@
+}
+
+interface Options {
+  dashcard?: BaseDashboardOrderedCard;
+  isEditing?: boolean;
+  onUpdateVisualizationSettings?: ({ text }: { text: string }) => void;
+  settings?: VisualizationSettings;
+}
+
+const defaultProps = {
+  dashcard: createMockDashboardCardWithVirtualCard(),
+  isEditing: false,
+  onUpdateVisualizationSettings: () => {
+    return;
+  },
+  settings: { text: "" },
+};
+
+const setup = (options: Options) => {
+  render(<Heading {...defaultProps} {...options} />);
+};
+
+describe("Text", () => {
+  describe("Saved (Not Editing)", () => {
+    it("should be able to render with text", () => {
+      const options = {
+        settings: getSettingsWithText("Example Heading"),
+      };
+      setup(options);
+
+      expect(
+        screen.getByTestId("saved-dashboard-heading-content"),
+      ).toHaveTextContent("Example Heading");
+    });
+  });
+
+  describe("Editing Dashboard", () => {
+    describe("Preview/Unfocused", () => {
+      it("should preview with placeholder and styling for no content", () => {
+        const options = {
+          settings: getSettingsWithText(""),
+          isEditing: true,
+        };
+        setup(options);
+
+        expect(
+          screen.getByTestId("editing-dashboard-heading-preview"),
+        ).toHaveTextContent("Heading");
+        expect(screen.getByTestId("editing-dashboard-heading-container"))
+          .toHaveStyle(`border: 1px solid ${color("brand")};
+                        color: ${color("text-light")};`);
+      });
+
+      it("should preview with text when it has content", () => {
+        const options = {
+          settings: getSettingsWithText("Example Heading"),
+          isEditing: true,
+        };
+        setup(options);
+
+        expect(
+          screen.getByTestId("editing-dashboard-heading-preview"),
+        ).toHaveTextContent("Example Heading");
+      });
+    });
+
+    describe("Edit/Focused", () => {
+      it("should focus input when clicked", () => {
+        const options = {
+          settings: getSettingsWithText(""),
+          isEditing: true,
+        };
+        setup(options);
+
+        screen.getByTestId("editing-dashboard-heading-preview").click();
+        expect(
+          screen.getByTestId("editing-dashboard-heading-input"),
+        ).toHaveFocus();
+      });
+
+      it("should have input placeholder when it has no content", () => {
+        const options = {
+          settings: getSettingsWithText(""),
+          isEditing: true,
+        };
+        setup(options);
+
+        screen.getByTestId("editing-dashboard-heading-preview").click();
+        expect(screen.getByPlaceholderText("Heading")).toBeInTheDocument();
+      });
+
+      it("should render input text when it has content", () => {
+        const options = {
+          settings: getSettingsWithText("Example Heading"),
+          isEditing: true,
+        };
+        setup(options);
+
+        screen.getByTestId("editing-dashboard-heading-preview").click();
```

Comment:
It's not clear to me whether the chained `.click()` here uses fireEvent or userEvent 🤔  

Where possible, we prefer to use the `userEvent` library for testing user interactions in user tests. It puts a slightly higher level of abstraction that, in many cases, more accurately simulates user interactions than firing raw events. 

https://testing-library.com/docs/user-event/intro

Reply:
Thank you for this!